rero_ils/modules/local_fields/api.py

```diff
@@ -0,0 +1,136 @@
+# -*- coding: utf-8 -*-
+#
+# RERO ILS
+# Copyright (C) 2020 RERO
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published by
+# the Free Software Foundation, version 3 of the License.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+"""API for manipulating local_fields."""
+
+from functools import partial
+
+from elasticsearch_dsl import Q
+
+from .models import LocalFieldIdentifier, LocalFieldMetadata
+from ..api import IlsRecord, IlsRecordsIndexer, IlsRecordsSearch
+from ..fetchers import id_fetcher
+from ..minters import id_minter
+from ..providers import Provider
+
+# provider
+LocalFieldProvider = type(
+    'LocalFieldProvider',
+    (Provider,),
+    dict(identifier=LocalFieldIdentifier, pid_type='lofi')
+)
+# minter
+local_field_id_minter = partial(id_minter, provider=LocalFieldProvider)
+# fetcher
+local_field_id_fetcher = partial(id_fetcher, provider=LocalFieldProvider)
+
+
+class LocalFieldsSearch(IlsRecordsSearch):
+    """LocalFieldsSearch."""
+
+    class Meta:
+        """Search only on local_field index."""
+
+        index = 'local_fields'
+        doc_types = None
+        fields = ('*', )
+        facets = {}
+
+        default_filter = None
+
+
+class LocalField(IlsRecord):
+    """LocalField class."""
+
+    minter = local_field_id_minter
+    fetcher = local_field_id_fetcher
+    provider = LocalFieldProvider
+    model_cls = LocalFieldMetadata
+
+    @classmethod
+    def create(cls, data, id_=None, delete_pid=False,
+               dbcommit=False, reindex=False, **kwargs):
+        """Create local_field record."""
+        record = super(LocalField, cls).create(
+            data, id_, delete_pid, dbcommit, reindex, **kwargs)
+        return record
+
+    def update(self, data, dbcommit=False, reindex=False):
+        """Update local_field record."""
+        super(LocalField, self).update(data, dbcommit, reindex)
+        return self
+
+    def extended_validation(self, **kwargs):
+        """Validate record against schema."""
+        # Check if local fields resource exists
+        from ..local_fields.api import LocalField
+
+        data = self.replace_refs()
+        parent_ref = data['parent_ref']
+        localFieldsRecord = LocalField.get_local_fields_by_resource(
+            parent_ref['type'], parent_ref['pid'], data['organisation']['pid']
+        )
+        if localFieldsRecord:
+            return False
+        return True
```

Comment:
I don't understand to purpose of this test ? Using `get_local_fields_by_resource` with parent_ref parameter, you will find all localfields linked to your parent (yourself include !). So this test should always return True.
Maybe you want to test if parent exists ? In this case the method `pids_exists_in_data` should be sufisant.

Reply:
This test is used to check the existence for an organization because the resource must be unique.